### ISSUE TITLE
Parsing error design and fancy errors.

### ISSUE THREAD

**Issue by Kixunil** [rust-bitcoin/rust-bitcoin#1406]
We started improving parsing errors but I'm not really sure what's the best design. There are two main things I'm concerned about: sensible carrying of the input and advanced error information such as spans or multiple errors.

The first issue is that some errors are nested. Recently proposed `FromHexStr` trait is an example of the situation where there are two checks being done, one of them being optional. So that begs the question where the input should go.

The second issue is I'd love to enable fancier errors - such as spanned error messages pluggable into `codespan-reporting` (I do **not** want to add it as a dep, just an example of consuming crate) and ability to report multiple errors. For the latter it'd be cool to report all positions of invalid hex chars.

Here's an example of beautiful error message from `configure_me_codegen` crate:

```
error: the identifier `a=bit**loong   and ###weird@parameter` contains invalid characters
   ┌─ unknown file:42:18
   │
42 │         name = "a=bit**loong   and ###weird@parameter"
   │                  ^   ^^     ^^^   ^^^^     ^ ... and this char
   │                  │   │      │     │         
   │                  │   │      │     ... and these chars
   │                  │   │      ... and these chars
   │                  │   ... and these chars
   │                  this char is invalid

"#;
```

I guess carrying around more error information should be feature-gated. I think this would require #1143 to some degree.

I'm thinking of something like this:

```rust
pub struct ParseError<E> {
    input: ParseInput,
    first_error: E,
    #[cfg(feature = "multi-error")]
    other_errors: Vec<E>,
}

impl<E> ParseError<E> {
    #[cfg(feature = "error-input")]
    pub fn input(&self) -> &str {
        &self.input
    }

    #[cfg(feature = "multi-error")]
    pub fn sources(&self) -> impl Iterator<Item=&E> + '_ {
        std::iter::once(&self.first_error).chain(&self.other_errors)
    }
}
```

Though it's still imperfect as we could for instance detect missing `0x` prefix in hex and invalid hex characters and there can only be one invalid prefix error. Also maybe we should have a single error for invalid characters just as in `configure_me` which only carries multiple spans within the same error.

Defining a completely new error for each item is a possibility, although a bit annoying. Maybe it'll be required anyway. In that case maybe we should have a trait instead:

```rust
pub trait ParseError {
    /// Iterator of labels
    type Labels: Iterator<Item=(String, Span)>; // GAT would be very useful here...
    /// Human-readable English explanation of the expected format.
    fn expecting(&self) -> &str;

    /// Additional hint relevant to the specific input.
    ///
    /// This is intended to explain more details about the specific error.
    fn hint(&self) -> Option<&str>;

    /// Returns an iterator of labels for characters within the input to explain the error better.
    fn labels(&self) -> Self::Labels;
}

pub struct Span {
    /// Starts from 0 which points to the beginning of the input passed to a parsing function
    pub start: usize,
    /// How many bytes are affected.
    pub len: usize,
}
```

Another can of worms is localization. I'm not sure if I want to deal with it but maybe we should keep the possibility open somehow.

**Comment by apoelstra** [comment]
I don't think we need to feature-gate any extra error info. If we have large errors we should box them and make sure that we have sensible accessors to get information out of them (and even to downcast).

We may need an alternate "dumb" representation when we have nostd, I suppose.

**Comment by Kixunil** [comment]
Yeah, that's a possibility too and probably not too bad. If someone is parsing strings instead of binary maybe they should switch the format as well. That being said, since `no_std` will likely need same gates as `multi-error` it might be useful to have it optional.

**Comment by apoelstra** [comment]
Yep, that makes sense -- I guess what I'm proposing is that we aggressively try to shrink the non-`multi-error` versions of our errors, even as we increase the size of the `multi-error` versions.

**Comment by sanket1729** [comment]
This seems like a big task. Would love to see concrete examples of rust-bitcoin data structures where a fancier display might be helpful. My only guess it is just amount? Everything else is just consensus-defined hex serialized strings.

**Comment by Kixunil** [comment]
@sanket1729 `Network`, `LockTime`... I could find more. Also hex errors could point to the locations of the invalid characters.
